File: Arduino/4004/ALU.cpp
#include "ALU.h"
#include "INST.h"

static TIMING *timing ;
static DATA *data ;
static byte acc = 0 ;               // The accumulator
static byte tmp = 0 ;               // The tmp register
static bool cy = 0 ;                // Carry
static byte ada = 0 ;               // Input A of the adder (4 bits)
static byte adb = 0 ;               // Input B of the adder (4 bits)
static bool adc = 0 ;               // Input C od the adder (1 bit)
static byte acc_out = 0 ;           // Accumulator output
static bool cy_out = 0 ;            // Carry output
static byte add = 0 ;               // The result of the adder (not a register in the real 4004 as the adder is combinational)


void ALU_reset(){
  acc = 0 ;
  tmp = 0 ;
  cy = 0 ;
  ada = 0 ;
  adb = 0 ;
  adc = 0 ;
  acc_out = 0 ;
  cy_out = 0 ;
  add = 0 ;
}


void ALU_setup(TIMING *t, DATA *d){
  timing = t ;
  data = d ;
  ALU_reset() ;
  ALU_timing() ;
}


void ALU_timing(){
  timing->M12([]{    // Initialize the ALU before each instruction
    ada = 0 ;
    tmp = 0xF ;
    adc = 0 ;
  }) ;
  
  timing->X12([]{    // Save acc and cy to their output registers. 
    acc_out = acc ;
    cy_out = cy ;
  }) ;
          
  timing->X21([]{    // Set the bus with the proper initialization value, depending on the current instruction.
    if (ope()){
      enableInitializer() ;
    }
  }) ;
          
  timing->X22clk1([]{  // Set input B by sampling data from the bus (n0342, for non IO instructions).
    if (! io()){
      tmp = data->read() ;
    }
  }) ;
  
  timing->X22clk2([]{  // Set input B by sampling data from the bus (n0342, for IO instructions).
    if (io()){
      tmp = data->read() ;
    }
  }) ;
}


// The Adder implementation
void runAdder(bool invertADB, bool saveAcc, bool saveCy, bool shiftL, bool shiftR){
  adb = tmp ;
  if (invertADB){
    adb = ~adb & 0xF ;
  }

  add = ada + adb + adc ;
  bool co = add >> 4 ;
  add = add & 0xF ;

  if (shiftL){
    cy = add >> 3 ;
    acc = (add << 1 | cy_out) & 0xF ;
  }    
  else if (shiftR){
    cy = add & 1 ;
    acc = cy_out << 3 | add >> 1 ;
  }
  else {
    if (saveAcc){
      acc = add ;
    }
    if (saveCy){
      // WARNING: a bit of DAA logic here...
      if (daa() && (cy_out || acc_out > 9)){
        cy = 1 ;
      }
      else {
        cy = co ;
      }
    }
  }
}


// Set input A
void setADA(bool invert){
  ada = acc ;
  if (invert){
    ada = ~ada & 0xF ;
  }
}


// Set input C
void setADC(bool invert, bool one){
  if (one){
    adc = 1 ;
  }
  else {
    adc = cy ;
    if (invert){
      adc = ~adc & 1 ;
    }
  }
}

      
// Place acc_out on the bus.
void enableAccOut(){
  data->write(acc_out) ;
}


// Place adder result on the bus
void enableAdd(){
  data->write(add) ;
}


// Place carry out on the bus
void enableCyOut(){
  data->write(cy_out) ;
}


// Bus initialization routine. This is really clever...
void enableInitializer(){
  byte d = 0 ;
  if (get_opa() >> 3){
    if (daa()){
      if (cy_out || acc_out > 9){
        d = 6 ;
      }
      else {
        d = 0 ;
      } 
    }
    else if (tcs()){
      d = 9 ;
    }
    else if (kbp()){
      if (acc_out == 4){
        d = 3 ;
      }
      else if (acc_out == 8){
        d = 4 ;
      }
      else if (acc_out > 2){
        d = 15 ;
      }
      else {
        d = acc_out ;
      }
    }
    else {
      d = 0xF ;
    } 
  }
  
  data->write(d) ;
}


// Is acc == 0?
bool accZero(){
  return acc_out == 0 ;
}


// Is adder result == 0?
bool addZero(){
  return add == 0 ;
}


// Is carry == 1?
bool carryOne(){
  return cy_out ;
}


byte getAcc(){
  return acc ; 
}

File: chips/modules/alu.py
from chips.modules.timing import *
from hdl import *


# Tihs class implemts the ALU of the 4004.


class alu:
    def __init__(self, inst, timing, data):
        self.data = data
        self.inst = inst
        self.inst.alu = self
        self.acc = 0                # The accumulator
        self.tmp = 0                # The tmp register
        self.cy = 0                 # Carry
        self.ada = 0                # Input A of the adder (4 bits)
        self.adb = 0                # Input B of the adder (4 bits)
        self.adc = 0                # Input C od the adder (1 bit)
        self.acc_out = 0            # Accumulator output
        self.cy_out = 0             # Carry output
        self.add = 0                # The result of the adder (not a register in the real 4004 as the adder is combinational)
 
        self.timing = timing

        @M12    # Initialize the ALU before each instruction
        def _():
            self.ada = 0
            self.tmp = 0xF
            self.adc = 0

        @X12    # Save acc and cy to their output registers. 
        def _():
            self.acc_out = self.acc
            self.cy_out = self.cy
            #print(self.timing.cycle, "acc_out", self.acc_out, "cy_out", self.cy_out)
        
        #@X21    # Set the bus with the proper initialization value, depending on the current instruction.
        #def _():
        #    if self.inst.ope():
        #        self.enableInitializer()
        
        @X22clk1  # Set input B by sampling data from the bus (n0342, for non IO instructions).
        def _():
            if not self.inst.io():
                self.tmp = self.data.v
                # print(self.timing.cycle, "tmp", self.tmp)

        @X22clk2  # Set input B by sampling data from the bus (n0342, for IO instructions).
        def _():
            if self.inst.io():
                self.tmp = self.data.v
                # print(self.timing.cycle, "tmp", self.tmp)

    # The Adder implementation
    def runAdder(self, invertADB=False, saveAcc=False, saveCy=False, shiftL=False, shiftR=False):
        self.adb = self.tmp
        if invertADB:
            self.adb = ~self.adb & 0xF

        self.add = self.ada + self.adb + self.adc
        co = self.add >> 4
        self.add = self.add & 0xF

        # print("acc:{} ada:{} tmp:{} adb:{} cy:{}, adc:{}".format(self.acc, self.ada, self.tmp, self.adb, self.cy, self.adc))
        # print("add:{} co:{}".format(self.add, co))

        if shiftL:
            self.cy = self.add >> 3
            self.acc = (self.add << 1 | self.cy_out) & 0xF
        elif shiftR:
            self.cy = self.add & 1
            self.acc = self.cy_out << 3 | self.add >> 1
        else:
            if saveAcc:
                self.acc = self.add
            if saveCy:
                # WARNING: a bit of DAA logic here...
                if self.inst.daa() and (self.cy_out or self.acc_out > 9):
                    self.cy = 1
                else:
                    self.cy = co
        #print(self.timing.cycle, self.inst.opr, self.inst.opa, "runAdder", self.add, self.cy)

    # Set input A
    def setADA(self, invert=False):
        self.ada = self.acc
        if invert:
            self.ada = ~self.ada & 0xF
        #print(self.timing.cycle, "setADA", self.ada)

    # Set input C
    def setADC(self, invert=False, one=False):
        if one:
            self.adc = 1
        else:
            self.adc = self.cy
            if invert:
                self.adc = ~self.adc & 1
        #print(self.timing.cycle, "setADC", self.adc)

    # Place acc_out on the bus.
    def enableAccOut(self):
        self.data.v = self.acc_out
        #print(self.timing.cycle, "enableAccOut", self.acc_out)

    # Place adder result on the bus
    def enableAdd(self):
        self.data.v = self.add
        #print(self.timing.cycle, "enableAdd", self.add)

    # Place carry out on the bus
    def enableCyOut(self):
        self.data.v = self.cy_out
        #print(self.timing.cycle, "enableCyOut", self.cy_out)

    # Bus initialisation routine. This is really clever...
    def enableInitializer(self):
        data = 0
        if self.inst.opa >> 3:
            if self.inst.daa():
                if self.cy_out or self.acc_out > 9:
                    data = 6
                else:
                    data = 0 
            elif self.inst.tcs():
                data = 9
            elif self.inst.kbp():
                if self.acc_out == 4:
                    data = 3
                elif self.acc_out == 8:
                    data = 4
                elif self.acc_out > 2:
                    data = 15
                else:
                    data = self.acc_out
            else:
                data = 0xF 
        #print(self.timing.cycle, "enableInitializer", data)        
        self.data.v = data

    # Is acc == 0?
    def accZero(self):
        #print(self.timing.cycle, "accZero", 1 if self.acc_out == 0 else 0)
        return 1 if self.acc_out == 0 else 0

    # Is adder result == 0?
    def addZero(self):
        #print(self.timing.cycle, "addZero", 1 if self.add == 0 else 0)
        return 1 if self.add == 0 else 0

    # Is carry == 1?
    def carryOne(self):
        #print(self.timing.cycle, "carryOne", self.cy_out)
        return self.cy_out

